### DISCUSSION TITLE
[Question] How Function for missing values

### DISCUSSION THREAD

**Discussion by avilca** [flet-dev/flet#2007] [Q&A] [answered]
### Question
I want to validate that my controls don't have empty fields with an AlertDialog, I've created a function that validates empty fields but still allows me to enter empty fields and doesn't load my AlertDialog.

Quiero validar que mis controles no tengan campos vacios  con un AlertDialog, e creado una funcion que valida campos vacios pero igual me sigue permitiendo ingresar campos vacios y no me carga mi AlertDialog,

Show code

### Code sample

```python
#CONTROLES (Controls)

    codpro = TextField(label="CÓDIGO PRODUCTO", width = 200)
    nompro = TextField(label="NOMBRE PRODUCTO", width = 600)

    #combo de categoria (Category)
    ddcate = Dropdown(label="CATEGORIA",
        width=150,
        options=[
            dropdown.Option("dulce"),
            dropdown.Option("salado"),
        ],
    )

    #combo de peso (weight)
    ddpeso = Dropdown(label="PESO",
        width=100,
        options=[
            dropdown.Option("KG"),
            ft.dropdown.Option("G"),
        ],
    )

    def mostrar_mensaje(e):
          page.dialog = dialog2
          dialog2.open = True
          page.update()


   def cerrar_mensaje(e):
         dialog2.open = False
         page.update()

    #DIALOG PARA MENSAJE DE CAMPOS VACIOS
    dialog2 = AlertDialog(
        title=Text("Algún campo esta vacío", bgcolor="red"),
        actions=[
            ElevatedButton("Aceptar",
                       on_click=cerrar_mensaje, 
                       bgcolor= "red", 
                       color= "white",
                      ),        
        ],
    )    



 #FUNCION PARA VALIDAR CAMPOS VACIOS (FUNCTION Empty Values)
    def valida_campos(e):

        lista = [codpro, nompro, ddcate, ddpeso]

        if len(lista) == 0:
            mostrar_mensaje(e)


#FUNCION para AGREGAR (Function Add)    
    def agregar(e):
        
        valida_campos(e)

        tablaproducto.rows.append(
            DataRow(
                cells=[
                    DataCell(Text(codpro.value)),
                    DataCell(Text(nompro.value)),
                    DataCell(Text(ddcate.value)),
                    DataCell(Text(ddpeso.value)),
                    DataCell(
                        Row([
                        IconButton("delete", 
                                   icon_color ="red",
                                   on_click = eliminar,
                                    ),
                        IconButton("create", 
                                   icon_color ="blue",
                                   on_click = editar
                                   ),    
                            ])
                        ),
                ]
            )
        )

#Boton AGREGAR
    BtnAgregar = ElevatedButton(
        text="Agregar", 
        bgcolor="green",
        color="white",
        on_click=agregar
        )
```

### ------------------------------------------------------

- [X] I have searched for answers to my question both in the issues and in previous discussions.

**Comment by ingdesarr1** [comment]
En que momento realiza la validación? Tiene un boton adicional para eso?

**Reply by avilca** [reply]
Si hay boton Agregar, acabo de modificar el codigo que puse al principio, poniendo la funcion agregar y el boton

**Reply by ingdesarr1** [reply]
Cuando presiona el boton agregar, salta a la funcion agregar, y en esta salta a la funcion "valida_campos". Esta funcion, realmente no esta validando si los espacios estan vacios o no, ya que solo agrego los objetos a una lista y luego pregunta si la lista es igual a 0, lo cual no es cierto, y por tanto, no abre el mensaje. Lo que deberia hacer es consultar por la propiedad value de cada uno de los campos y validar si estan vacios o si no tienen texto. Si en alguno de esos casos se cumple, entonces que abra el mensaje, si no, que ingrese a la tabla y guarde los datos.

**Reply by avilca** [reply]
Intente poniendo el If not pero aún asi me sigue permitiendo ingrersar dato vacios

I tried putting the If not but it still allows me to enter empty data.

`
def agregar(e):
        
        
        lista = [codpro, nompro, ddcate, ddpeso]
        
    
  
        if not lista:
            mostrar_mensaje(e) 
        else:
        #   agregar 
            tablaproducto.rows.append(
                DataRow(
                    cells=[
                        DataCell(Text(codpro.value)),
                        DataCell(Text(nompro.value)),
                        DataCell(Text(ddcate.value)),
                        DataCell(Text(ddpeso.value)),
                        DataCell(
                            Row([
                            IconButton("delete", 
                                    icon_color ="red",
                                    on_click = eliminar,
                                        ),
                            IconButton("create", 
                                    icon_color ="blue",
                                    on_click = editar
                                    ),    
                                ])
                            ),
                    ]
                )
            )`

Alguna sugerencia o posible solución ??

Any suggestions or possible solutions??

**Reply by ingdesarr1** [reply, marked as answer]
el if deberia ser algo asi:

```
check = False
for campo in lista:
    if campo.value == '' or not campo.value:
        print("Faltan datos")
        check = True

if check == False:
    ##guardar los datos en la tabla##
```

**Reply by avilca** [reply]
Sii me funciono  pero basta que ingrese el campo codigo y los demas campos los dejo vacío igual permite ingresar datos en blanco. Ahora cuando dejo el campo codigo vacío y los demas campos los lleno si valida, te muestro el código y la imagen

`
def agregar(e):
        
        
        lista = [codpro, nompro, ddcate, ddpeso]
        
        check = False

        for campo in lista:
            if campo.value == '' or not campo.value:
                mostrar_mensaje(e)
                check = True

            if check == False:
                    
            #   agregar 
                tablaproducto.rows.append(
                    DataRow(
                        cells=[
                            DataCell(Text(codpro.value)),
                            DataCell(Text(nompro.value)),
                            DataCell(Text(ddcate.value)),
                            DataCell(Text(ddpeso.value)),
                            DataCell(
                                Row([
                                IconButton("delete", 
                                        icon_color ="red",
                                        on_click = eliminar,
                                            ),
                                IconButton("create", 
                                        icon_color ="blue",
                                        on_click = editar
                                        ),    
                                    ])
                                ),
                        ]
                    )
                )
`

![Captura desde 2023-11-02 14-40-32](https://github.com/flet-dev/flet/assets/7683412/41ef3ba8-f870-47c8-b1e4-4484d4fb26f9)

**Reply by ingdesarr1** [reply]
En ese caso fue por un error de espacios, cuando escribi el codigo no puso los espacios correctamente. La condición del check == False debe ser despues de terminar el for, para que termine de validar todas las entradas y luego si abra el mensaje.

**Reply by avilca** [reply]
Muchas gracias @ingdesarr1

**Reply by avilca** [reply]
Una ultima pregunta que es lo que hace la variable check ?, es como un validador o verificador ??

**Reply by ingdesarr1** [reply]
Si. La variable check es solo una bandera que utilizo para saber si algun campo esta vacio o no. Se inicializa como Falso, pensando que todos los campos estan llenos, pero al momento de recorrer el for, si alguno de los campos esta vacio, se pone en True, lo que evita que los datos se agreguen.

**Reply by avilca** [reply]
Gracias @ingdesarr1 por la explicación
